Direct Messages

kterry1 opened the pull request:
### What does this pull request aim to achieve? 💯
_Please be concise and link any related issue(s):_
closes https://github.com/FightPandemics/FightPandemics/issues/1457
closes https://github.com/FightPandemics/FightPandemics/issues/1455
closes https://github.com/FightPandemics/FightPandemics/issues/1453
closes https://github.com/FightPandemics/FightPandemics/issues/1440
closes https://github.com/FightPandemics/FightPandemics/issues/1454
closes https://github.com/FightPandemics/FightPandemics/issues/1599
closes https://github.com/FightPandemics/FightPandemics/issues/1865

<!--### 🚨Before submitting this pull request🚨:-->

<!--_Please do **NOT** submit this PR if you have not done the following:_-->

- [X] I have checked that no one else is working on similar changes.
- [X] I have read the latest [**README**](README.md) and understand the development workflow.
- [X] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix. The branch is created in a cloned version of this repo, **not a forked repo**.
- [X] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [X] There are no merge conflicts.
- [X] There are no console warnings and errors.
- [X] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".

kterry1 committed:
Added private message modal

kterry1 committed:
Removed extra curly brace

kterry1 committed:
Added mobile view

kterry1 committed:
Added Mobile View for Success Message

joshmorel commented:
> This pull request is NOT complete and is NOT ready for review. 

Cool. It's hard to find the button (under reviewers list) but in such a I usually switch to draft (already done for you):

![image](https://user-images.githubusercontent.com/10546720/91744584-b780b800-eb87-11ea-8b36-73a4f961b7a0.png)

kterry1 committed:
Added inbox page

kterry1 commented:
As I see that Draft option now @joshmorel Thanks!

kterry1 committed:
original post and toggle post button

kterry1 committed:
Added messaging bubles

kterry1 committed:
Fixed spacing issues with message bubbles

kterry1 committed:
Added input box border

kterry1 committed:
Redirected users to login page when not logged in

kterry1 committed:
Added 'No Messages' screen and mobile view for private messages

kterry1 commented:
Still in progress.

kterry1 committed:
Fixed Success Modal Button placement

kterry1 committed:
Allowed toggle between current chat and chat list in mobile view

kterry1 committed:
Disabled submit button for Modal and Inbox when no text has been typed

kterry1 committed:
Added failed to send modal and cleaned up code a bit

kterry1 committed:
Redesigned Recipient's Orginal Post bubble

kterry1 committed:
Refactored Inbox components and redesigned message bubbles

kterry1 committed:
Replaced px values with respective em values

kterry1 committed:
Refactored Inbox.js for performance improvements

kterry1 committed:
Added useRef for input box focus

kterry1 committed:
Testing different comments for message modal

kterry1 committed:
Moved styled components outside of functional ones

Zaydme committed:
It's an absolute mess, but you can send messages!

Zaydme commented on backend/config.js:
```diff
@@ -79,6 +81,15 @@ const config = {
+    options: {
+      pingInterval: 10000,
+    },
+    redis: {
+      host: configData.REDIS_HOST,
+      port: configData.REDIS_PORT
+    }
```
We need Redis in testing env

Zaydme committed:
messaging functionality update, ± cleanup

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
realtime users status update

Zaydme committed:
added post reference if message sent from feed

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
turn urls into hyperlinks in messages, + user status bug fix

Zaydme committed:
removed debugging leftover, and yes I use console.log to debug 😳

Zaydme committed:
optimized how things work

Zaydme committed:
don't emit JOIN_ROOM if already in same room

Zaydme committed:
load old messages (just prototype), ± mobile compatibility

mannykary committed:
Add Redis backend environment variables

Zaydme committed:
Merge pull request #1580 from FightPandemics/feature/1579-add-redis

#1579 Add Redis backend environment variables

Zaydme committed:
change socket url to work in review env

Zaydme committed:
changing socket url based on env

Zaydme committed:
add socket to nginx config

Zaydme committed:
messages grouping and timestamps

Zaydme committed:
made linter ignore some lines

Zaydme committed:
correctly add the unread mark when recipient is offline

Zaydme committed:
added avatar to reripient header

Zaydme committed:
fixed success modal visual bug

Zaydme committed:
updated status badge, unread messages indicator now shows numbers

Zaydme committed:
fixed mistake

Zaydme committed:
clear chatLog when leaving all rooms.

Zaydme committed:
styling to match design

Zaydme committed:
fixed user status state

Zaydme committed:
updated message modal to match design, fixed status state once again (hopefully)

Zaydme committed:
updated some icons to match design

Zaydme committed:
various changes

Zaydme committed:
header inbox icon with unread count

Zaydme committed:
removed unread count from the blue dot indicator

Zaydme committed:
show empty imbox screen if the user had no room, or no room selected

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
inbox shortcut with unread badge for mobile side Nav

Zaydme committed:
real-time message notifications

Zaydme committed:
don't auto join first room

Zaydme committed:
fixed a bug where socket leaves it's own room and stops receiving notifications

Zaydme committed:
display number of unread threads and not total number of unread messages

Zaydme committed:
updated message grouping timeframe, and timestamp format

Zaydme committed:
messages deleting

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Messages editing

Zaydme committed:
added the (edited) message indicator

Zaydme committed:
various improvements, code cleanup, and bug fixes

Zaydme committed:
eslint-disable-line react-hooks/exhaustive-deps

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
fix missing comma

Zaydme committed:
expandable textarea with shift&enter or text overflow

Zaydme committed:
mobile layout improvements

Zaydme committed:
prevent virtual keyboard from opening on mobile

mannykary committed:
Add Redis container

Zaydme committed:
updated socket backend to work with clustering

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Added Redis installation guide to Readme

Zaydme committed:
'Block' and 'archive' front-end

Zaydme committed:
user blocking/unblocking backend

Zaydme committed:
archiving + messages requests!

Zaydme committed:
leave All rooms when leaving inbox page to keep receiving messages notifications

Zaydme committed:
Styling edited and deleted messages

Zaydme committed:
added message button on profile page.

Zaydme committed:
fixed a bug where user is not displayed on messages list after sending message from feed

Monica0843 commented:
layout on mobile : iPhone7+Safari: 1. in message center - long message is cut off and not scrollable to allow user to read the whole massge. Second message after that gets cut off too.  2. When sending a new message from Post - Message icon. There is only a "View Message" icon on Mobile, while there is a "Done" icon on Desktop(not sure this is an issue)

![image (1)](https://user-images.githubusercontent.com/48136524/96059123-190c9580-0e5b-11eb-871c-4c7331fd3930.png)
![image (2)](https://user-images.githubusercontent.com/48136524/96059124-19a52c00-0e5b-11eb-9029-cd34ad8eac7b.png)
![image (3)](https://user-images.githubusercontent.com/48136524/96059126-1a3dc280-0e5b-11eb-9f91-8cc31e2f20ea.png)

Monica0843 commented:
Macbook+Chrome: when the message "To" receipt has a really long name the "View message" icon on "Your message was successfully sent" has layout issue 
![image](https://user-images.githubusercontent.com/48136524/96058777-283f1380-0e5a-11eb-93e8-501cf79814af.png)

Zaydme commented:
@Monica0843 please use https://github.com/FightPandemics/FightPandemics/issues/1599 for Feedback 👍

Zaydme committed:
Various bug fixes

Zaydme committed:
visual style fixes

Zaydme committed:
fixed message modal avatar & user cannot message their own orgs

Zaydme committed:
don't scroll down when clicking "load more"

Zaydme committed:
Fixed message modal styling

Zaydme committed:
inbox style fixes

Zaydme committed:
regex infinite loop fix

Zaydme committed:
Fixed edit control out of screen

Zaydme committed:
JWT authentication for websocket

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
various bug fixes

Zaydme committed:
better mobile message editing + characters counter

Zaydme committed:
keep user on inbox page after refresh

Zaydme committed:
update thread name on user update

Zaydme committed:
fixed ios overlapping bug

Zaydme committed:
Header inbox icon for mobile

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
fixed previous commit causing issue

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
remove duplicate lines

Zaydme committed:
removed duplicate key created for messageModal and caused comment button to break.

Zaydme committed:
use middleware to check socket identification

Zaydme committed:
clear messageModal textarea on send success

Zaydme committed:
fixed some styling issues

Zaydme committed:
(for web notifications) send notification to all the user's browsers if many are open (part1)

wadehrarshpreet committed:
style: fix message modal design

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
#1448 - Add search bar on help board with a search button

 - Add tabs search
 - Make sure the index is created for testing
 - Preserve  and use regex not  search if location is defined
 - Keyword highlighting
 - Change visible HELP_TYPE filters depending on tab
 - Don't refetch posts if no filters were selected
 - Do not reset page if no filters were selected
 - Don't ignore stop words + set minimum keyword length
 - Add search mobile version
 - Fixed search safari + brave browsers visual glitches
 - Removed some filters on mobile from Poeple/Orgs to match desktop, and fixed some unnecessary post refetch
 - Remove default language from User/Org index
 - For full $text search
 - Don't submit search if keyword too short
 - Auto hide search bar on tablets and small screens

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

jessica-lau committed:
Add Fixture Files for Cypress Tests (#1674)

* added fixture files and updated sign in spec

* fixed cypress.json file

* added additional fixtures and updated tests

* added additional fixtures and updated tests

sujatha-m committed:
#1663 - Modified Quit filters button to Clear filter button (#1668)

* Modified Quit filters button to Clear filter button

* Moved the fix from en_GB.json to en_US.json

jessica-lau committed:
Cypress e2e - Add cookie banner/logo POM files and refactor tests (#1706)

* added cookie banner and logo POM files and refactored e2e tests

* refactored additional tests

wadehrarshpreet committed:
improvement: Chat UI fixes

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
improvement: settings chat UI fix

emanuelquerty committed:
#614 - check email not verified (#1703)

* Fixes internal server error message on login with email not verified

* Makes the comment more concise

wadehrarshpreet committed:
improvement: alertbox action button UI fix

wadehrarshpreet committed:
improvement: long touch menu options chat

Zaydme committed:
Revert '#1448 - Add search bar on help board with a search button' This reverts commit 0df102b186287aa9475d7a0c2adb213b3b92880a.

Zaydme committed:
fixed unread indicator position

Zaydme committed:
internationalization of messaging feature

Zaydme committed:
adjusted bubble border radius

Zaydme committed:
fixed typos en_US (messaging)

wadehrarshpreet committed:
fix: message ui fix left sidebar & other issue

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
fix: message request issue fix

Zaydme committed:
ignored threads turn back into requests on new message

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

wadehrarshpreet committed:
improvement: confirm dialog added for ignore & block & minor ui fixes

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
fix: unarchive screen & message modal ui fix(mobile)

mannykary committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
backend code cleanup

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
improvement: updated profile page UI

Zaydme committed:
Added GTM for messaging

Zaydme committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
fix: minor style fix

wadehrarshpreet committed:
Merge branch 'feature/message-modal/1457' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

Zaydme committed:
Small UI fixes

wadehrarshpreet committed:
Merge branch 'staging' of https://github.com/FightPandemics/FightPandemics into feature/message-modal/1457

wadehrarshpreet committed:
fix: profile ui fix

wadehrarshpreet committed:
improvement: updated org profile view

Zaydme committed:
fix: newly signed up users not getting connected to WS

Zaydme committed:
mobile: hide chat list and show current chat when clicking "View Message"

Zaydme committed:
Updated messaging models indexes

Zaydme committed:
fixed bug: user can send message to pending conversation from helpboard before accepting it.

Zaydme committed:
Message button on Poeple/Orgs cards

Zaydme committed:
fixed small bug caused by previous commit

wadehrarshpreet committed:
fix: feed color in myprofile

mannykary committed:
Add indexes and emailSent field

Zaydme committed:
removed unused indexes

Zaydme committed:
Combined some duplicated socket events into one

Zaydme committed:
Fixed user status not updating in realtime

Zaydme committed:
messaging dates internationalizations (global moment.js internationalization)

Zaydme committed:
UI fixes and general UX improvements

Zaydme committed:
made dropdowns trigger on click only

Zaydme committed:
auto focus editing textarea

mannykary committed:
Refine emailSent flipping logic

mannykary committed:
Merge pull request #1822 from FightPandemics/feature/1741-update-message-thread-schemas

#1741 update message thread schemas

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Merge branch 'staging' into feature/message-modal/1457

Zaydme committed:
Messaging as Org

Zaydme committed:
Fix mobile blank page after unarchiving from settings

Zaydme committed:
Fix unarchive button not visible

Zaydme committed:
Hide contextMenu from deleted messages

Zaydme committed:
Fix scrollbar style in messageModal

Zaydme committed:
Prevent mobile keyboard from bouncing (close/open)

Zaydme committed:
Fix "Help board" button on post page when post not found

Zaydme committed:
fix inbox header icon not showing correct number

Zaydme committed:
Now you can't block someone before they accept your request

Zaydme committed:
Fix messageModal button hover style

Zaydme committed:
organization profile UI fixes

Zaydme committed:
message modal disabled button styling

Zaydme committed:
update org avatar in thread participant ref

ddveloper commented on backend/.env.example_local:
```diff
@@ -12,4 +12,4 @@ AIRTABLE_API_KEY=TODO
 AIRTABLE_BASE_ID=appKRi7YYHy6CBar8
 SENDGRID_API_KEY=TODO
 SENDGRID_CONTACTS_LIST_ID=TODO
-
```
add these one? (just saw we have these default value in config) do we still need these?
```
REDIS_HOST=localhost
REDIS_PORT=6379
```

Zaydme replied:
Yes this is for Redis (elasticache) which is now needed for socket.io to work with aws load balancing.

ddveloper commented on backend/.env.example_local:
```diff
@@ -12,4 +12,4 @@ AIRTABLE_API_KEY=TODO
 SENDGRID_API_KEY=TODO
 SENDGRID_CONTACTS_LIST_ID=TODO
-
+LOGGER_LEVEL=debug
```
not sure if this is needed

Zaydme replied:
Just added that for socket.io debugging. So we we don't have a lot of unnecessary logs in PROD. But since this is just .env.example anyone can change it to what they want.

ddveloper commented on backend/lib/models/Thread.js:
```diff
@@ -3,64 +3,71 @@ const { Schema, model, ObjectId } = require("mongoose");
 
 const CONVERSATION_STATUS_OPTIONS = [
-  "accepted", "archived", "blocked", "pending"
+  "accepted",
```
I think it's good to add some comment here, to clarify that we have 3 buttons for the user to push, `accept`, `ignore` and `block`, I guess if user press nothing, it's in `pending` state
Also another question is if `pending` and `ignore` has same logic?

Zaydme replied:
pending, is shown to the user as a message request. But if they clicked ignore, it becomes ignored, which means it's there in the db but not show to the user, until they receive a new message, it becomes pending again.

ddveloper commented on backend/lib/models/Thread.js:
```diff
@@ -3,64 +3,71 @@ const { Schema, model, ObjectId } = require("mongoose");
+  },
+  photo: String,
+  status: {
+    default: "accepted",
```
will this `status` be `accepted` even user doesn't press the `accept` button?

Zaydme replied:
For the sender it is accepted by default, for the recipient it will be set to pending.

mannykary commented on client/src/components/Inbox/EmptyInbox.js:
```diff
@@ -0,0 +1,83 @@
+      <img
+        className="empty-inbox-logo"
+        src={emptyinbox}
+        alt="Empty Inbox Page"
```
Should alt attributes be localized?

Zaydme replied:
I don't think so, I have never seen them localized anywhere

ddveloper commented on client/src/components/Feed/Post.js:
```diff
@@ -497,7 +498,13 @@ const Post = ({
-        postId={postId}
+        postAuthorName={post.author.name}
+        postAuthorAvatar={post.author.photo}
+        isOwnPost={
```
just a reminder this logic could be changed later, see #1843 for details

Zaydme replied:
Yes, that will probably need to be changed depending on what PR gets merged first

ddveloper commented on client/src/components/Header/Header.js:
```diff
@@ -150,14 +154,44 @@ const Header = ({
   onOrganisationChange,
   navSearch,
   setOrganisationId,
+  ws,
```
change to a better name?

Zaydme replied:
renamed to webSocket :+1:

ddveloper commented on client/src/components/Inbox/CurrentChat.js:
```diff
@@ -0,0 +1,131 @@
+        threadId: room._id,
+      });
+    setEditingMessageId(null);
+  }, [getChatLog, room, setEditingMessageId]);
```
need `// eslint-disable-next-line react-hooks/exhaustive-deps` to prevent lint mdofication

ddveloper commented on client/src/components/Inbox/InputBox.js:
```diff
@@ -0,0 +1,372 @@
+      window.screen.width >= parseInt(mq.phone.wide.maxWidth)
+    )
+      inputRef.current.focus();
+  }, [inputRef]);
```
need a `// eslint-disable-next-line react-hooks/exhaustive-deps`

ddveloper commented on client/src/components/Inbox/InputBox.js:
```diff
@@ -0,0 +1,372 @@
+      return setInputExpanded(true);
+    if ((text.match(/\n/g) || []).length) setInputExpanded(true);
+    else setInputExpanded(false);
+  }, [inputExpanded, inputRef, setInputExpanded, text]);
```
need a `// eslint-disable-next-line react-hooks/exhaustive-deps`

ddveloper commented on client/src/components/Inbox/Messages.js:
```diff
@@ -0,0 +1,374 @@
+
+  React.useEffect(() => {
+    getScrollToBottom.current = scrollToBottom;
+  }, [getScrollToBottom, scrollToBottom]);
```
// eslint-disable-next-line react-hooks/exhaustive-deps

ddveloper commented on client/src/components/Inbox/Messages.js:
```diff
@@ -0,0 +1,374 @@
+        const relativeTime = getRelativeTime(message.createdAt);
+        return (
+          <>
+            {message.authorId !== (user._id || user.id) ? (
```
do we need to check organisation case here? or use the `actorId` later?

Zaydme replied:
"user" is actually the "actor" here.
The actor was passed as user to those components in `pages/Inbox.js`

ddveloper commented on client/src/components/Profile/ProfileComponents.js:
```diff
@@ -68,174 +64,168 @@ export const CreatePostDiv = styled.div`
+
 export const ProfileLayout = styled.div`
-  background-color: ${colors.lighterGray};
+  background-color: #fbfbfd;
```
this may be add into the `colors` group

ddveloper commented on client/src/components/Profile/ProfileComponents.js:
```diff
@@ -254,15 +244,12 @@ export const UserInfoDesktop = styled.div`
+  margin-top: 2.4rem;
+  font-size: 1.4rem;
+  line-height: 2rem;
+  color: #282828;
```
this may need to check the `colors`

ddveloper commented on client/src/pages/Inbox.js:
```diff
@@ -0,0 +1,197 @@
+    } catch (e) {
+      // e
+    }
+  }, []);
```
// eslint-disable-next-line react-hooks/exhaustive-deps

ddveloper commented on client/src/pages/Inbox.js:
```diff
@@ -0,0 +1,197 @@
+      getChatLog({
+        threadId: room._id,
+      });
+  }, [getChatLog, room]);
```
// eslint-disable-next-line react-hooks/exhaustive-deps

ddveloper commented on client/src/templates/layouts/Main.js:
```diff
@@ -7,12 +7,12 @@ const Main = styled.main`
   display: flex;
   align-items: inherit;
   justify-content: center;
+  ${(props) => `${props.isProfile ? "background: #f5f5f9;" : ""}`}
```
add into the `colors`

ddveloper commented on client/src/components/Feed/MessagesModal/MessageModal.js:
```diff
@@ -0,0 +1,241 @@
+            isFromUserCard={isFromUserCard}
+            id={gtmId}
+          >
+            <img src={activeemail} />
```
do we need a `alt=`

ddveloper commented on client/src/components/Feed/MessagesModal/MessageModal.js:
```diff
@@ -0,0 +1,241 @@
+            isFromUserCard={isFromUserCard}
+            id={gtmId}
+          >
+            <img src={activeemail} />
```
do we need a `alt=`?

ddveloper commented on backend/lib/plugins/socket.js:
```diff
@@ -0,0 +1,449 @@
+    io.use(cookieParser());
+    // customHook, to run a request on every node
+    // every socket.io server executes below code, when customRequest is called
+    io.of("/").adapter.customHook = (request, cb) => {
```
change `cb` to `callBack`?

ddveloper reviewed (commented):
I didn't review the css part

Zaydme committed:
UI improvements + requested changes

Zaydme committed:
UI fixes related to account switch

rabbi08 reviewed (approved):
This features is working fine. Thanks

mannykary committed:
Add LB stickiness for Socket.IO

mannykary committed:
Add second instance in staging

mannykary reviewed (approved):
LGTM
